teps/0137-cloudevents-controller.md

```diff
@@ -0,0 +1,527 @@
+---
+status: proposed
+title: CloudEvents controller
+creation-date: '2023-06-19'
+last-updated: '2023-06-19'
+authors:
+- '@afrittoli'
+collaborators: []
+---
+
+# TEP-0137: CloudEvents controller
+---
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Tekton supports [sending CloudEvents][tekton-ce-docs] for events associated
+with `TaskRuns`, `PipelineRuns` and `CustomRuns`. Events associated with the
+latter are generated by a dedicated controller binary, while the former two
+are generated by the `TaskRun` and `PipelineRun` controller respectively.
+
+In this TEP we propose to move all the CloudEvents generation logic to the
+separate controller. The new controller will support new versions and types
+of events:
+
+- Tekton v2 events, which embed `v1.TaskRun` and `v1.PipelineRun` as opposed
```

Comment:
Trying to understand this kind of version mapping, is it possible to keep the same version across the tekton resources and events?

Reply:
Does this version of events has any influence from the CDEvents project?

Reply:
About aligning versions with API, the initial version of the events was `v1`, so I'm afraid it won't be possible.
Moving forward, I hope Tekton events will eventually be deprecated and we will stick with CDEvents.

Reply:
CDEvents has its own [versioning scheme](https://cdevents.dev/docs/primer/#versioning), when new versions of CDEvents become available, Tekton may update its CDEvents support to newer versions. I don't think we should support multiple versions of CDEvents at once.